### PR TITLE
Reduce stack size in division

### PR DESCRIPTION
**Pull request by kaidokert**
Removes 2 extra temporary full number copies from
division #15

## Summary by Sourcery

Replace the previous division routine with an optimized in-place algorithm that uses bit-level helper methods to eliminate extra temporaries and reduce stack usage; update division and remainder operators to leverage the new implementation and add comprehensive tests covering helper functions and division correctness.

New Features:
- Add helper methods for bit-level operations: set_bit, cmp_shifted, get_shifted_word, sub_shifted, and sub_with_borrow.
- Introduce div_assign_impl to perform in-place division on a mutable dividend and return the remainder.

Enhancements:
- Simplify div_impl to forward to div_assign_impl on a copy.
- Update DivAssign and RemAssign operator implementations to call the in-place division routine instead of reallocating temporaries.

Tests:
- Add extensive tests for the optimized division implementation, helper methods, shifted operations, operator forwarding, stack efficiency, and division property-based checks.

## Summary by CodeRabbit

* **New Features**
  * Improved division performance with a new, optimized in-place algorithm for large unsigned integers.
  * Added new helper methods for advanced bitwise operations.

* **Bug Fixes**
  * Enhanced correctness and reliability of division and remainder operations, especially for edge cases.

* **Tests**
  * Introduced extensive new tests to verify division accuracy and helper method behavior.

### CODE REVIEW

File: src/fixeduint.rs [hunk 1]
```diff
@@ -387,38 +387,188 @@ impl<T: MachineWord, const N: usize> FixedUInt<T, N> {
...
+            // Perform subtraction with borrow
+            let (result, new_borrow) =
+                self.sub_with_borrow(self.array[i], other_shifted_word, borrow);
+            self.array[i] = result;
+            borrow = if new_borrow { T::one() } else { T::zero() };
+        }
+    }
+
+    /// Subtract with borrow: a - b - borrow, returns (result, did_borrow)
+    fn sub_with_borrow(&self, a: T, b: T, borrow: T) -> (T, bool) {
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/fixeduint.rs]: done

File: src/fixeduint/mul_div_impl.rs [hunk 2]
```diff
@@ -249,7 +249,7 @@ impl<T: MachineWord, const N: usize> core::ops::RemAssign<Self> for FixedUInt<T,
         if other.is_zero() {
             maybe_panic(PanicReason::RemByZero)
         }
-        *self = self.div_rem(&other).1;
+        *self = Self::div_assign_impl(self, &other);
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/fixeduint/mul_div_impl.rs]: done

File: src/fixeduint/mul_div_impl.rs [hunk 3]
```diff
@@ -190,7 +190,7 @@
         if other.is_zero() {
             maybe_panic(PanicReason::DivByZero)
         }
-        *self = Self::div_impl(self, other);
+        Self::div_assign_impl(self, other);
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/fixeduint/mul_div_impl.rs:193]: done
